File: src/types/FamiliarsTrick.ts
import * as DB from "tsondb/schema/dsl"
import { OldParameter } from "./_ActivatableSkill.js"
import { IndefiniteOneTimeCost } from "./_ActivatableSkillCost.js"
import {
  DurationUnitValue,
  FixedDuration,
  IndefiniteDuration,
} from "./_ActivatableSkillDuration.js"
import { AnimalTypeIdentifier, PropertyIdentifier } from "./_Identifier.js"
import { ResponsiveText, ResponsiveTextOptional } from "./_ResponsiveText.js"
import { NestedTranslationMap } from "./Locale.js"
import { Errata } from "./source/_Erratum.js"
import { src } from "./source/_PublicationRef.js"

export const FamiliarsTrick = DB.Entity(import.meta.url, {
  name: "FamiliarsTrick",
  namePlural: "FamiliarsTricks",
  displayName: "Familiar’s Trick",
  displayNamePlural: "Familiar’s Tricks",
  type: () =>
    DB.Object({
      animal_types: DB.Required({
        comment: `The animal types this trick is available to. Either it is available to all or only a list of specific animal types.

If no animal types are given, the animal disease applies to all animal types.`,
        type: DB.Array(AnimalTypeIdentifier(), { uniqueItems: true }),
      }),
      parameters: DB.Required({
        comment: "Measurable parameters of a familiar’s trick.",
        type: DB.IncludeIdentifier(FamiliarsTrickPerformanceParameters),
      }),
      property: DB.Required({
        comment: "The associated property.",
        type: DB.IncludeIdentifier(FamiliarsTrickProperty),
      }),
      ap_value: DB.Optional({
        comment:
          "The AP value the familiar has to pay for. It may also be that a specific is known by all familiar by default. In the latter case the field is not set.",
        type: DB.Integer({ minimum: 1 }),
      }),
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "FamiliarsTrick",
        DB.Object({
          name: DB.Required({
            comment: "The familiar’s trick’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          effect: DB.Required({
            comment: "The effect description.",
            type: DB.String({ minLength: 1, isMarkdown: true }),
          }),
          cost: DB.Optional({
            isDeprecated: true,
            type: DB.IncludeIdentifier(OldParameter),
          }),
          duration: DB.Optional({
            isDeprecated: true,
            type: DB.IncludeIdentifier(OldParameter),
          }),
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

const FamiliarsTrickProperty = DB.Enum(import.meta.url, {
  name: "FamiliarsTrickProperty",
  values: () => ({
    Fixed: DB.EnumCase({ type: PropertyIdentifier() }),
    Indefinite: DB.EnumCase({ type: DB.IncludeIdentifier(IndefiniteFamiliarsTrickProperty) }),
  }),
})

const IndefiniteFamiliarsTrickProperty = DB.TypeAlias(import.meta.url, {
  name: "IndefiniteFamiliarsTrickProperty",
  type: () =>
    DB.Object({
      translations: NestedTranslationMap(
        DB.Optional,
        "IndefiniteFamiliarsTrickProperty",
        DB.Object({
          description: DB.Required({
            comment: "A description of the property.",
            type: DB.IncludeIdentifier(ResponsiveText),
          }),
        }),
      ),
    }),
})

const FamiliarsTrickPerformanceParameters = DB.Enum(import.meta.url, {
  name: "FamiliarsTrickPerformanceParameters",
  comment: "Measurable parameters of a familiar’s trick.",
  values: () => ({
    OneTime: DB.EnumCase({
      type: DB.IncludeIdentifier(FamiliarsTrickOneTimePerformanceParameters),
    }),
    OneTimeInterval: DB.EnumCase({
      type: DB.IncludeIdentifier(FamiliarsTrickOneTimeIntervalPerformanceParameters),
    }),
    Sustained: DB.EnumCase({
      type: DB.IncludeIdentifier(FamiliarsTrickSustainedPerformanceParameters),
    }),
  }),
})

const FamiliarsTrickOneTimePerformanceParameters = DB.TypeAlias(import.meta.url, {
  name: "FamiliarsTrickOneTimePerformanceParameters",
  type: () =>
    DB.Object({
      cost: DB.Required({
        type: DB.IncludeIdentifier(FamiliarsTrickOneTimeCost),
      }),
      duration: DB.Required({
        type: DB.IncludeIdentifier(FamiliarsTrickOneTimeDuration),
      }),
    }),
})

const FamiliarsTrickOneTimeCost = DB.Enum(import.meta.url, {
  name: "FamiliarsTrickOneTimeCost",
  values: () => ({
    Fixed: DB.EnumCase({ type: DB.IncludeIdentifier(FamiliarsTrickFixedOneTimeCost) }),
    All: DB.EnumCase({ type: DB.IncludeIdentifier(FamiliarsTrickAllOneTimeCost) }),
    Indefinite: DB.EnumCase({ type: DB.IncludeIdentifier(IndefiniteOneTimeCost) }),
  }),
})

const FamiliarsTrickFixedOneTimeCost = DB.TypeAlias(import.meta.url, {
  name: "FamiliarsTrickFixedOneTimeCost",
  type: () =>
    DB.Object({
      ae_value: DB.Required({
        comment: "The AE cost value.",
        type: DB.Integer({ minimum: 1 }),
      }),
      lp_value: DB.Optional({
        comment: "The LP cost value.",
        type: DB.Integer({ minimum: 1 }),
      }),
      interval: DB.Optional({
        comment: "The interval in which you have to pay the AE cost again.",
        type: DB.IncludeIdentifier(DurationUnitValue),
      }),
      translations: NestedTranslationMap(
        DB.Optional,
        "FamiliarsTrickFixedOneTimeCost",
        DB.Object(
          {
            per: DB.Optional({
              comment:
                "The cost have to be per a specific countable entity, e.g. `8 KP per person`.",
              type: DB.IncludeIdentifier(ResponsiveTextOptional),
            }),
          },
          { minProperties: 1 },
        ),
      ),
    }),
})

const FamiliarsTrickAllOneTimeCost = DB.TypeAlias(import.meta.url, {
  name: "FamiliarsTrickAllOneTimeCost",
  type: () =>
    DB.Object({
      minimum: DB.Optional({
        comment: "The minimum AE the familiar has to have/spend.",
        type: DB.Integer({ minimum: 1 }),
      }),
    }),
})

const FamiliarsTrickOneTimeDuration = DB.Enum(import.meta.url, {
  name: "FamiliarsTrickOneTimeDuration",
  values: () => ({
    Immediate: DB.EnumCase({ type: null }),
    Fixed: DB.EnumCase({ type: DB.IncludeIdentifier(FixedDuration) }),
    Indefinite: DB.EnumCase({ type: DB.IncludeIdentifier(IndefiniteDuration) }),
  }),
})

const FamiliarsTrickOneTimeIntervalPerformanceParameters = DB.TypeAlias(import.meta.url, {
  name: "FamiliarsTrickOneTimeIntervalPerformanceParameters",
  type: () =>
    DB.Object({
      cost: DB.Required({
        type: DB.IncludeIdentifier(FamiliarsTrickOneTimeIntervalCost),
      }),
    }),
})

const FamiliarsTrickOneTimeIntervalCost = DB.TypeAlias(import.meta.url, {
  name: "FamiliarsTrickOneTimeIntervalCost",
  type: () =>
    DB.Object({
      ae_value: DB.Required({
        comment: "The AE cost value.",
        type: DB.Integer({ minimum: 1 }),
      }),
      lp_value: DB.Optional({
        comment: "The LP cost value.",
        type: DB.Integer({ minimum: 1 }),
      }),
      interval: DB.Required({
        comment: "The duration granted/added by paying the given AE cost.",
        type: DB.IncludeIdentifier(DurationUnitValue),
      }),
    }),
})

const FamiliarsTrickSustainedPerformanceParameters = DB.TypeAlias(import.meta.url, {
  name: "FamiliarsTrickSustainedPerformanceParameters",
  type: () =>
    DB.Object({
      cost: DB.Required({
        type: DB.IncludeIdentifier(FamiliarsTrickSustainedCost),
      }),
    }),
})

const FamiliarsTrickSustainedCost = DB.TypeAlias(import.meta.url, {
  name: "FamiliarsTrickSustainedCost",
  type: () =>
    DB.Object({
      ae_value: DB.Required({
        comment: "The AE cost value.",
        type: DB.Integer({ minimum: 1 }),
      }),
      lp_value: DB.Optional({
        comment: "The LP cost value.",
        type: DB.Integer({ minimum: 1 }),
      }),
      interval: DB.Optional({
        comment: "The interval in which you have to pay the AE cost again.",
        type: DB.IncludeIdentifier(DurationUnitValue),
      }),
    }),
})

File: Sources/OptolithDatabaseSchema/Types/FamiliarsTrick.swift
import FileDB

@Model
public struct FamiliarsTrick {
    /// The animal types this trick is available to. Either it is available to all or only a list of specific animal types.
    ///
    /// If no animal types are given, the animal disease applies to all animal types.
    @UniqueItems
    @Relationship(AnimalType.self)
    let animal_types: [AnimalType.ID]

    /// Measurable parameters of a familiar’s trick.
    let parameters: FamiliarsTrickPerformanceParameters

    /// The associated property.
    let property: FamiliarsTrickProperty

    /// The AP value the familiar has to pay for. It may also be that a specific is known by all familiar by default. In the latter case the field is not set.
    @Minimum(1)
    let ap_value: Int?

    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation {  // FamiliarsTrickTranslation
        /// The familiar’s trick’s name.
        @MinLength(1)
        let name: String

        /// The effect description.
        @MinLength(1)
        @Markdown
        let effect: String

        @available(*, deprecated, message: "Use language-independent performance parameters instead")
        let cost: OldParameter

        @available(*, deprecated, message: "Use language-independent performance parameters instead")
        let duration: OldParameter

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }
}

@ModelEnum
public enum FamiliarsTrickProperty {
    case fixed(PropertyIdentifierObject)
    case indefinite(IndefiniteFamiliarsTrickProperty)
}

@Embedded
public struct IndefiniteFamiliarsTrickProperty {
    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]?

    struct Translation {  // IndefiniteFamiliarsTrickPropertyTranslation
        /// A description of the property.
        let description: ResponsiveText
    }
}

/// Measurable parameters of a familiar’s trick.
@ModelEnum
public enum FamiliarsTrickPerformanceParameters {
    case oneTime(FamiliarsTrickOneTimePerformanceParameters)
    case oneTimeInterval(FamiliarsTrickOneTimeIntervalPerformanceParameters)
    case sustained(FamiliarsTrickSustainedPerformanceParameters)
}

@Embedded
public struct FamiliarsTrickOneTimePerformanceParameters {
    let cost: FamiliarsTrickOneTimeCost

    let duration: FamiliarsTrickOneTimeDuration
}

@ModelEnum
public enum FamiliarsTrickOneTimeCost {
    case fixed(FamiliarsTrickFixedOneTimeCost)
    case all(FamiliarsTrickAllOneTimeCost)
    case indefinite(IndefiniteOneTimeCost)
}

@Embedded
public struct FamiliarsTrickFixedOneTimeCost {
    /// The AE cost value.
    @Minimum(1)
    let ae_value: Int

    /// The LP cost value.
    @Minimum(1)
    let lp_value: Int?

    /// The interval in which you have to pay the AE cost again.
    let interval: DurationUnitValue?

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]?

    @Embedded
    @MinProperties(1)
    struct Translation {  // FamiliarsTrickFixedOneTimeCostTranslation
        /// The cost have to be per a specific countable entity, e.g. `8 KP per person`.
        let per: ResponsiveTextOptional?
    }
}

@Embedded
public struct FamiliarsTrickAllOneTimeCost {
    /// The minimum AE the familiar has to have/spend.
    @Minimum(1)
    let minimum: Int?
}

@ModelEnum
public enum FamiliarsTrickOneTimeDuration {
    case immediate
    case fixed(FixedDuration)
    case indefinite(IndefiniteDuration)
}

@Embedded
public struct FamiliarsTrickOneTimeIntervalPerformanceParameters {
    let cost: FamiliarsTrickOneTimeIntervalCost
}

@Embedded
public struct FamiliarsTrickOneTimeIntervalCost {
    /// The AE cost value.
    @Minimum(1)
    let ae_value: Int

    /// The LP cost value.
    @Minimum(1)
    let lp_value: Int?

    /// The duration granted/added by paying the given AE cost.
    let interval: DurationUnitValue
}

@Embedded
public struct FamiliarsTrickSustainedPerformanceParameters {
    let cost: FamiliarsTrickSustainedCost
}

@Embedded
public struct FamiliarsTrickSustainedCost {
    /// The AE cost value.
    @Minimum(1)
    let ae_value: Int

    /// The LP cost value.
    @Minimum(1)
    let lp_value: Int?

    /// The interval in which you have to pay the AE cost again.
    let interval: DurationUnitValue?
}
